File: diamond/python/src/diamond/diamond.py
def print_diamond(letter: str) -> str:
    if len(letter) != 1:
        raise ValueError("letter must be a single A-Z character")
    normalized = letter.upper()
    if not ("A" <= normalized <= "Z"):
        raise ValueError("letter must be a single A-Z character")

    n = ord(normalized) - ord("A")
    rows: list[str] = []
    for r in range(2 * n + 1):
        offset = r if r <= n else 2 * n - r
        rows.append(_build_row(offset, n))
    return "\n".join(rows)


def _build_row(offset: int, n: int) -> str:
    row_letter = chr(ord("A") + offset)
    leading = " " * (n - offset)
    if offset == 0:
        return leading + row_letter
    inner = " " * (2 * offset - 1)
    return leading + row_letter + inner + row_letter

File: diamond/csharp/src/Diamond/Diamond.cs
using System.Text;

namespace Diamond;

public static class Diamond
{
    public static string Print(char letter)
    {
        var normalized = char.ToUpperInvariant(letter);
        if (normalized < 'A' || normalized > 'Z')
        {
            throw new ArgumentException(
                "letter must be a single A-Z character",
                nameof(letter));
        }

        var n = normalized - 'A';
        var rows = new List<string>(2 * n + 1);
        for (var r = 0; r <= 2 * n; r++)
        {
            var offset = r <= n ? r : 2 * n - r;
            rows.Add(BuildRow(offset, n));
        }
        return string.Join("\n", rows);
    }

    private static string BuildRow(int offset, int n)
    {
        var leading = n - offset;
        var row = new StringBuilder(2 * n + 1);
        row.Append(' ', leading);
        var rowLetter = (char)('A' + offset);
        row.Append(rowLetter);
        if (offset > 0)
        {
            row.Append(' ', 2 * offset - 1);
            row.Append(rowLetter);
        }
        return row.ToString();
    }
}
